### PR TITLE
Add Buffer.takeHead

### PR DESCRIPTION
**Pull request by e5l**

### CODE REVIEW

File: src/commonMain/kotlin/io/ktor/io/Buffer.kt [hunk 1]
```diff
@@ -265,6 +265,18 @@ public interface Buffer : Closeable {
...
+     * Split this [Buffer] into head and tail.
+     *
+     * The head buffer will have bytes between 0 and index inside. And the tail buffer will have bytes between [index]
+     * and [capacity]. The [readIndex] and [writeIndex] will be pointing to the same bytes in the result buffers.
+     *
+     * The head of split is returned and this buffer becomes the tail.
+     *
+     * @return head of split
+     */
+    public fun takeHead(index: Int = writeIndex): Buffer
```
Comment (whyoleg) [inline src/commonMain/kotlin/io/ktor/io/Buffer.kt:278]: TBH, I'm not sure about the naming and overall it's contract.

Now we have:
1. absolute primitive operations (get/set)
2. relative primitive operations (read/write)
3. copy bulk(array/buffer) operations (both absolute and relative)

And this `takeHead`, which DETACHES HEAD part of current buffer from 0 up to some index looks not so good...
I see, that `netty5` have the same operation called `split`

Do you have some usecases for it's usage. I have some ideas, but it's hard for now to understand if it's will be useful.
BTW, why we (and netty) do return HEAD and not TAIL as result of this operation?

P.S. without experimenting with it it's hard for me to understand what is better

**Review by whyoleg — CHANGES_REQUESTED** [review]

**Comment by ptitjes** [comment]
@e5l I don't know if this is intended (and if it is, I fail to understand why) but:

- in `ByteArrayBuffer.takeHead`, you manually create a `ByteArrayBuffer` (and pass it the same referenceCounter and pool) and in `ByteArrayBuffer.close` you recycle `this` to the pool, whereas
- in `JvmBuffer.takeHead`, you manually create a `JvmBuffer` (and pass it the same referenceCounter, pool, and this as `origin`) and in `JvmBuffer.close` you recycle the `origin` buffer (or this, if null).

Shouldn't you also pass an origin buffer in the case of `ByteArrayBuffer`? Otherwise, won't you be adding to the pool a buffer that is sharing a `ByteArray` with a buffer instance that points to the same `ByteArray`? Maybe, there is something I am missing here...)

If we should also have an `origin` in `ByteArrayBuffer`, then I agree with @whyoleg that `ReferenceCounter` could hold the pointers to the `origin` buffer and the `pool`.

**Comment by ptitjes** [comment]
@e5l Maybe I said something wrong. I searched for a scenario where that would happen. What I came with is:

```
val pool = ByteArrayBufferPool()
val originBuffer = pool.borrow()
val otherBuffer = originBuffer.takeHead(64)
originBuffer.close()
otherBuffer.close()
```

However, it just happens that the `otherBuffer` will go back to the pool instead of the `originBuffer`. But `originBuffer` can never go back to the pool, so I guess this is not a problem?

File: src/commonMain/kotlin/io/ktor/io/ByteArrayBuffer.kt [hunk 2]
```diff
@@ -1,31 +1,71 @@
...
+import io.ktor.io.internals.*
+import io.ktor.io.internals.ReferenceCounter
+import kotlin.math.max
 import kotlin.math.min
 
 public const val DEFAULT_POOL_CAPACITY: Int = 2000
 public const val DEFAULT_BUFFER_SIZE: Int = 1024 * 16
 
-public class ByteArrayBuffer(
+public class ByteArrayBuffer internal constructor(
```
[the remark this answers is not in the document]
Comment (e5l) [reply src/commonMain/kotlin/io/ktor/io/ByteArrayBuffer.kt:11]: let me check

File: src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        allocations += allocation
+        return allocation.instance
+    }
+
+    override fun dispose() {
+        checkLeaks()
+    }
+
+    override fun recycle(instance: ByteArrayBuffer) {
+        val allocation = allocations.find { it.instance === instance }
```
[the remark this answers is not in the document]
Comment (e5l) [reply src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt:22]: Thanks, will fix

File: src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        val leaks = allocations.filter { !it.isDisposed }
+        if (leaks.isEmpty()) return
+
+        val stacks = leaks.joinToString("\n") { it.stack }
+        throw IllegalStateException("Leaked ${allocations.size} objects: $stacks")
+    }
+}
+
+private class Allocation<T>(val instance: T) {
+    var stack: String = Exception().stackTraceToString()
```
[the remark this answers is not in the document]
Comment (e5l) [reply src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt:38]: The previous implementation was reusing instances, now it's not - will drop

File: src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt [hunk 5]
```diff
@@ -0,0 +1,56 @@
+package io.ktor.io.utils
+
+import io.ktor.io.*
+import kotlinx.atomicfu.atomic
+
+class LeakDetectingPool : ObjectPool<ByteArrayBuffer> {
+    private val allocations = mutableListOf<Allocation<ByteArrayBuffer>>()
```
[the remark this answers is not in the document]
Comment (e5l) [reply src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt:7]: Could you tell me if there are some libs in common?

File: src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+    override val capacity: Int = DEFAULT_POOL_CAPACITY
+
+    override fun borrow(): ByteArrayBuffer {
+        val array = ByteArray(DEFAULT_BUFFER_SIZE)
```
[the remark this answers is not in the document]
Comment (e5l) [reply src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt:11]: We can, but I'm not sure why it's needed in tests

File: src/jvmMain/kotlin/io/ktor/io/JvmBuffer.kt [hunk 7]
```diff
@@ -38,23 +44,23 @@ public class JvmBuffer(
      * All modifications of the [ByteBuffer] is reflected by the [JvmBuffer] itself.
      */
     @Suppress("CanBePrimaryConstructorProperty")
-    public var buffer: ByteBuffer = buffer
+    public var raw: ByteBuffer = view
```
[the remark this answers is not in the document]
Comment (e5l) [reply src/jvmMain/kotlin/io/ktor/io/JvmBuffer.kt:47]: and`underlying` should be private

File: src/jvmMain/kotlin/io/ktor/io/JvmBuffer.kt [hunk 8]
```diff
@@ -1,34 +1,40 @@
...
  * @constructor creates buffer prepared for reading.
  */
-public class JvmBuffer(
-    buffer: ByteBuffer,
-    private val pool: ObjectPool<ByteBuffer> = ByteBufferPool.Default
+public class JvmBuffer internal constructor(
+    view: ByteBuffer,
+    private val referenceCounter: ReferenceCounter,
+    private val pool: ObjectPool<JvmBuffer>,
+    private val origin: JvmBuffer? = null
```
[the remark this answers is not in the document]
Comment (e5l) [reply src/jvmMain/kotlin/io/ktor/io/JvmBuffer.kt:16]: I'm also thinking about `Disposable` like class, will experiment with that

**Comment by e5l** [comment]
@ptitjes yep it should be as in JvmBuffer, will add tests and fix

File: src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt (hunk 5 above)
[the remark this answers is not in the document]
Comment (whyoleg) [reply src/commonTest/kotlin/io/ktor/io/utils/LeakDetectingPool.kt:7]: have no idea
there is an issue to keep track on IdentityMap in youtrack: https://youtrack.jetbrains.com/issue/KT-8435/IdentityMap
But looks like we can easily implement it with single expect/actual for identity hash code (similar to what we had in old ktor-io)
Or while we still have `Allocation` object, we can just use `Set` with identity hash code in it implementation

File: src/jvmMain/kotlin/io/ktor/io/JvmBuffer.kt (hunk 8 above)
[the remark this answers is not in the document]
Comment (whyoleg) [reply src/jvmMain/kotlin/io/ktor/io/JvmBuffer.kt:16]: I also have some idea about pooling/refcounting/cleaner API. I will try to experiment a bit with it, and will give some feedback

**Comment by e5l** [comment]
Let me think a bit about different approach

**Pull request closed without merging**